c/enc/block_splitter.c

```diff
@@ -73,6 +77,273 @@ static void CopyLiteralsToByteArray(const Command* cmds,
   }
 }
 
+void BrotliSplitBlockCommandsFromStored(
+                        MemoryManager* m,
+                        const Command* cmds,
+                        const size_t num_commands,
+                        const size_t pos,
+                        const size_t mask,
+                        BlockSplit* cmd_split,
+                        const BlockSplitFromDecoder* cmd_split_decoder,
+                        size_t* cur_block_decoder) {
+    BROTLI_ENSURE_CAPACITY(
+        m, uint8_t, cmd_split->types, cmd_split->types_alloc_size,
+        cmd_split_decoder->num_blocks);
+    BROTLI_ENSURE_CAPACITY(
+        m, uint32_t, cmd_split->lengths, cmd_split->lengths_alloc_size,
+        cmd_split_decoder->num_blocks);
+    size_t cur_pos = pos;
+    cmd_split->num_blocks = 0;
+    cmd_split->num_types = 0;
+    size_t cur_length = 0;
+    /* Mapping of the types from decoder (they increase with each metablock)
+       to the appropriate types */
+    int* types_mapping = (int*)malloc(sizeof(int) * cmd_split_decoder->num_types);
+    for (int i = 0; i < cmd_split_decoder->num_types; ++i) {
+      types_mapping[i] = -1;
+    }
+    for (int i = 0; i < num_commands; ++i) {
+      const Command cmd = cmds[i];
+      /* If current command lies after the current block
+         that means we've finished the current block.
+         If some commands have fallen inside current block before
+         then need to save it*/
+      if (cur_pos >= cmd_split_decoder->positions_end[*cur_block_decoder] &&
+            cur_length > 0) {
+        /* If we haven't seen this decoder block type before
+          then save a mapping of it to a current num_types */
+        if (types_mapping[cmd_split_decoder->types[*cur_block_decoder]] == -1) {
+          types_mapping[cmd_split_decoder->types[*cur_block_decoder]] =
+                                                          cmd_split->num_types;
+        }
+        /* If the same block type as for a previous block then merge them */
+        if (cmd_split->num_blocks > 0 &&
+            types_mapping[cmd_split_decoder->types[*cur_block_decoder]] ==
+              cmd_split->types[cmd_split->num_blocks - 1]) {
+            cmd_split->lengths[cmd_split->num_blocks - 1] += cur_length;
+        } else {
+          /* Map the decoder block type to an appropriate type */
+          cmd_split->types[cmd_split->num_blocks] =
+                    types_mapping[cmd_split_decoder->types[*cur_block_decoder]];
+          cmd_split->lengths[cmd_split->num_blocks] = cur_length;
+          cmd_split->num_types = BROTLI_MAX(uint8_t, cmd_split->num_types,
+                                  cmd_split->types[cmd_split->num_blocks] + 1);
+          cmd_split->num_blocks++;
+        }
+        cur_length = 0;
+        (*cur_block_decoder)++;
+      }
+      /* Go through decoder blocks until a block with cur_pos inside is found */
+      while (cur_pos >= cmd_split_decoder->positions_end[*cur_block_decoder]) {
```

Comment:
Is the position of blocks incrementing?

Reply:
Yes, the position of block from decoder is cmd_split_decoder->positions_end[*cur_block_decoder] and it increases as cur_block_decoder increases. cur_pos is a position of current symbol